src/player.cpp

```diff
@@ -8569,16 +8552,67 @@ item::reload_option player::select_ammo( const item &base, const std::vector<ite
                 }
             }
         }
-        if( hotkey == -1 && last == ammo.typeId() ) {
-            // if this is the first occurrence of the most recently used type of ammo and the hotkey
-            // was not already set above then set it to the keypress that opened this prompt
-            hotkey = inp_mngr.get_previously_pressed_key();
-            last = std::string();
+        if( last == ammo.typeId() ) {
+            if( !last_key_bound && hotkey == -1 ) {
+                // If this is the first occurrence of the most recently used type of ammo and the hotkey
+                // was not already set above then set it to the keypress that opened this prompt
+                hotkey = last_key;
+                last_key_bound = true;
+            }
+            if( !last_key_bound ) {
+                // Pressing the last key defaults to the first entry of compatible type
+                default_to = i;
+                last_key_bound = true;
+            }
+        }
+        if( hotkey == last_key ) {
+            last_key_bound = true;
+            // Prevent the default from being used: key is bound to something already
+            default_to = -1;
         }
 
         menu.addentry( i, true, hotkey, draw_row( i ) );
     }
 
+    struct reload_callback : public uimenu_callback {
+        public:
+            std::vector<item::reload_option> &opts;
+            const std::function<std::string( int )> &draw_row;
+            int last_key;
+            int default_to;
+
+            reload_callback( std::vector<item::reload_option> &_opts,
+                             const std::function<std::string( int )> &_draw_row,
+                             int _last_key, int _default_to ) :
+                           opts( _opts ), draw_row( _draw_row ),
+                           last_key( _last_key ), default_to( _default_to )
+            {}
+
+            bool key( const input_event &event, int idx, uimenu * menu ) override {
+                auto &sel = opts[ idx ];
```

Comment:
This needs rearrnging, uimenu::query() calls this with idx = -1, causing an out of bounds access.
It looks straightforward so I'm just going to fix it up while merging.